docs/content/sips/017-service-chaining.md

```diff
@@ -0,0 +1,106 @@
+
+The preferred user experience is option 2: `http://spin.internal/<component-name>`
+
+**DETAIL NEEDED:** How will Spin request headers be set in this case?  One option is that Spin won't set them _at all_ and the calling component has to inject them if the receiver needs them.  Otherwise e.g. Consider a request to `http://spin.internal/accounts/314`.  `spin-path-info` (the 'wildcard match' bit of the path) is probably `/314` (bit after component name), but what about `spin-matched-route` or `spin-component-route`?  Are they just not set?  What of `spin-client-addr` - forwarded or skipped?
+
+**DETAIL NEEDED:** Should Spin automatically set an additional header indicating that this is a chained (unrouted) call?  If so, should it contain any additional data?  Or is it up to the caller to decide what, if anything, they want to disclose?
+
+## Permissions
+
+The developer must enable chaining by adding `http://spin.internal` to the list of allowed outbound hosts.  (We cannot use the application flag as permission because not all components may be allowed to make self-requests.)  We could and arguably should consider that any component with `self` permission also has chaining permission: if a component is allowed to make requests to other parts of the application, who cares if that's routed or chained?  However, the idea of private endpoints complicates this a bit - is there a situation in which a component should be allowed to access public routes but not private endpoints?
+
+**DECISION REQUIRED:** Should `allowed_outbound_hosts` including _either_ of `self` or `spin.internal` be enough to grant permission for self-requests using chaining?
+
+Currently, `self` allows access to all routes in the application.  The initial version of service chaining will be similar; if someone needs to restrict access to specific components, that can be a future option.
+
+# Runtime and framework considerations
+
+## Chaining requirements in the lock file
+
+The lock file format built into existing hosts is designed for extension, and so [avoids using serde "deny unknown fields"](https://github.com/fermyon/spin/blob/663593d1423ed3518744d6f797e6a3970575d617/crates/locked-app/src/locked.rs#L16).  Unfortunately, this means there is no way to add a prerequisites section to the lock file in such a way that existing hosts will reject it if present.  For example, if someone deploys a chaining-required app onto an older version of the containerd shim, it will appear to accept it, but will not chain.
+
+Unfortunately, I don't think there's anything we can do about this, except to introduce a new `spin_lock_version`.  Doing this naively would mean existing hosts would reject the lock file _whether or not it contains prerequisites_.  We could get smart and emit a v.next lock file _only_ if the app expressed prerequisites but hoo boy.
```

Comment:
Pragmatically, I think we could still get away with this since the locked app JSON format is only exposed externally in OCI artifacts which are not (yet) in common use outside of Fermyon-maintained systems. We could also improve this problem for the future by adding `deny_unknown_fields` to the top-level `LockedApp` schema (but not `metadata`; forcing careful thought about this kind of compat problem is very much the point there).

Reply:
It's exposed to external triggers: if we allow non-HTTP triggers to chain, then triggers linked against downlevel versions of the trigger crate need to reject applications with the chaining requirement.

Reply:
Another option would be to "get smart and emit a v.next lock file _only_ if the app expressed prerequisites" but drop that logic in a future major spin release.

Reply:
An option to avoid having to bump the version for similar purposes in the future:

- Update the `spin_lock_version` to `1`
- Add a new top-level field, e.g. `spin_lock_features = ["host_prereqs"]`; any features in this set must be known-supported by host
- Set `host_prereqs` iff there are any `required` host prereqs